src/lib.rs

```diff
@@ -216,73 +216,81 @@ macro_rules! scoped_thread_local {
+                        c.replace(Some(unsafe { cast_to_static(t) }))
+                    });
+                    let _reset = Reset { key: self.inner, val: prev };
+                    f()
                 }
 
-                // This wrapper helps to ensure that the 'static lifetime is not visible
-                // to the safe code.
-                fn cast_from_static<'a, 'b>(x: &'a Hkt<'static>) -> Hkt<'b> where 'a: 'b {
-                    *x
+                pub fn with<F, R>(&'static self, f: F) -> R
+                    where F: FnOnce(Hkt<'_>) -> R
+                {
+                    let val = self.inner.with(|c| c.get());
+                    let val = val.expect("cannot access a scoped thread local variable without calling `set` first");
+
+                    // This also asserts that Hkt is covariant
+                    f(cast_from_static(&val))
                 }
 
-                impl $name<'static> {
-                    pub fn set<F, R>(&'static self, t: Hkt<'_>, f: F) -> R
-                        where F: FnOnce() -> R
-                    {
-                        struct Reset {
-                            key: &'static LocalKey<Cell<Option<Hkt<'static>>>>,
-                            val: Option<Hkt<'static>>,
-                        }
-                        impl Drop for Reset {
-                            fn drop(&mut self) {
-                                self.key.with(|c| c.set(self.val.take()));
-                            }
-                        }
-                        let prev = self.inner.with(|c| {
-                            // Safety: we are only changing the lifetime. We enforce the
-                            // lifetime constraints via the `Reset` struct.
-                            c.replace(Some(unsafe { cast_to_static(t) }))
-                        });
-                        let _reset = Reset { key: self.inner, val: prev };
-                        f()
-                    }
+                /// Test whether this TLS key has been `set` for the current thread.
+                pub fn is_set(&'static self) -> bool {
+                    self.inner.with(|c| c.get().is_some())
+                }
+            }
+        };
 
-                    pub fn with<F, R>(&'static self, f: F) -> R
-                        where F: FnOnce(Hkt<'_>) -> R
-                    {
-                        let val = self.inner.with(|c| c.get());
-                        let val = val.expect("cannot access a scoped thread local variable without calling `set` first");
+        $(#[$attrs])*
+        $vis static $name: $name<'static> = {
+            type Hkt<$lt> = $ty;
+            use ::std::cell::Cell;
+            use ::std::option::Option;
 
-                        // This also asserts that Hkt is covariant
-                        f(cast_from_static(&val))
-                    }
+            thread_local!(static FOO: Cell<Option<Hkt<'static>>> = {
```

Comment:
If we're moving the thread local here, shouldn't the previous thread local be removed?

Reply:
Indeed, thanks for spotting that. That was my intention.
Fixed.